File: 0200-number-of-islands/0200-number-of-islands.java
class Solution {
    public int numIslands(char[][] grid) {
        int count=0;
        for(int i=0;i<grid.length;i++){
            for(int j=0;j<grid[0].length;j++){
                if(grid[i][j] == '1'){
                    count+=1;
                    dfs(i, j, grid);
                }
            }
        }
        return count;
    }
    void dfs(int row, int col, char[][] grid){
        if(row < 0 || row >= grid.length || col < 0 || col >= grid[0].length || grid[row][col] != '1'){
            return ;
        }
        grid[row][col] = '*';

        dfs(row-1, col, grid);
        dfs(row+1, col, grid);
        dfs(row, col-1, grid);
        dfs(row, col+1, grid);
    }
}

File: 0200-number-of-islands/0200-number-of-islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        m = len(grid)
        n = len(grid[0])

        visited = set()
        directions = [(-1, 0), (1, 0), (0, -1), (0, 1)]
        def bfs(ir, ic):
            dq = deque([(ir, ic)])
            
            while dq:
                nr, nc = dq.popleft()
                for r, c in directions:
                    row = nr+r
                    col = nc+c

                    if (row < 0 or row >= m or col < 0 or col >= n
                        or (row, col) in visited or grid[row][col] != "1"):
                        continue

                    dq.append((row, col))
                    visited.add((row, col))

                    grid[row][col] = "-1"

        count = 0
        for i in range(m):
            for j in range(n):
                if grid[i][j] == "1":
                    count += 1
                    bfs(i, j)
        return count

                    